tests: add pay test over unannounced channels

JssDWt opened the pull request:
This test fails with cln v24.08.2. Add this test, so it doesn't happen again.

## Checklist
Before submitting the PR, ensure the following tasks are completed. If an item is not applicable to your PR, please mark it as checked:

- [x] The changelog has been updated in the relevant commit(s) according to the [guidelines](https://docs.corelightning.org/docs/coding-style-guidelines#changelog-entries-in-commit-messages).
- [x] Tests have been added or modified to reflect the changes.
- [x] Documentation has been reviewed and updated as needed.
- [x] Related issues have been listed and linked, including any that this PR closes.

vincenzopalazzo commented on tests/test_pay.py:
```diff
@@ -6702,3 +6702,19 @@ def get_local_channel_by_id(node, chanid):
+    node_factory.join_nodes([l2, r1], wait_for_announce=True)
+
+    invoice = l3.rpc.invoice(2100000, 'receive', 'receive')
+    l1.rpc.pay(invoice['bolt11'])
```
I would preferer to just not blindly trust this by avoiding to check the result.

JssDWt replied:
Updated

vincenzopalazzo reviewed (commented):
left a comment, otherwise looks good to me

Concept ACK aa7a2a3baf46b0624a80f3dd9d6932789f853cf8

JssDWt commented:
The test was flaky
Added another random public node to ensure the route hint is added. 
Added a check that the recipient sees the public channels.
Added a check that the payment completed.
Rebased to resolve a conflict.

vincenzopalazzo reviewed (approved):
ACK 9b3c14313de8cc0003099fcc3036771b0f0faeca

vincenzopalazzo commented on tests/test_pay.py:
```diff
@@ -6731,3 +6731,26 @@ def test_fetchinvoice_with_payer_metadata(node_factory, bitcoind):
+    wait_for(lambda: len(l3.rpc.listchannels(destination=r2.info['id'])['channels']) > 0)
+
+    inv = l3.rpc.invoice(2100000, 'receive', 'receive')
+    l1.rpc.pay(inv['bolt11'])
+    wait_for(lambda: l1.rpc.listpays(inv['bolt11'])['pays'][0]['status'] == 'complete')
```
```suggestion
    result = l1.rpc.pay(inv['bolt11'])
    assert result["status"] == "complete", f"pay result is {result}"
```

This could also work, without adding `wait_for`, we should save some CPU cicly

vincenzopalazzo reviewed (commented):
I just rember that pay si returning the status too, so probably there is a clean way to check

vincenzopalazzo commented on tests/test_pay.py:
```diff
@@ -6731,3 +6731,26 @@ def test_fetchinvoice_with_payer_metadata(node_factory, bitcoind):
+    node_factory.join_nodes([l2, r2], wait_for_announce=True)
+
+    # Ensure recipient sees the channels
+    wait_for(lambda: len(l3.rpc.listchannels(destination=r1.info['id'])['channels']) > 0)
+    wait_for(lambda: len(l3.rpc.listchannels(destination=r2.info['id'])['channels']) > 0)
```
```suggestion
    wait_for(lambda: len(l3.rpc.listpeerchannels(destination=r1.info['id'])) > 0)
    wait_for(lambda: len(l3.rpc.listpeerchannels(destination=r2.info['id'])) > 0)
```

probably here, it is better to use the new API, the same also CPU here and have inside the test a better example of how to perform these operations

JssDWt replied:
`l3` doesn't have a channel with `r1` and `r2`, but this is here in order to ensure the public channels of `l2` are in the graph of `l3`.  Because `l2` needs public channels, otherwise `l3` will not add a route hint in the invoice.

rustyrussell replied:
Note: you can use the "dev-routes" parameter to `invoice` (in developer mode) to write any routehints you want?

JssDWt replied:
updated to use dev-routes

vincenzopalazzo reviewed (commented):
sorry for the floppy review, adding another consideration

vincenzopalazzo reviewed (approved):
ACK https://github.com/ElementsProject/lightning/pull/7844/commits/abd22011ff8e7ebab9bb92d4ab8d5e6f78fe2103
